### PR TITLE
feat(scheduling): implement azure, cinder, ebs and gce as filter plugin

### LINKED ISSUE #86613: Move volume limits predicates to their Filter plugins [Migration Phase 2]
/sig scheduling
/priority important-soon

Part of #85822

/assign @draveness

### PR DESCRIPTION
**Pull request by draveness**
/sig scheduling
/kind feature
/priority important-soon

Fixes #86613 

1. Implement azure, cinder, ebs and gce as filter plugin
2. Initialize cloud-specific volume limits plugin with NonCSILimits
3. Remove related predicates

We could remove the `non_csi.go` and `non_csi_test.go` after the CSI migration graduates to GA in https://github.com/kubernetes/kubernetes/issues/82375

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+	FilterVolume           func(vol *v1.Volume) (id string, relevant bool)
+	FilterPersistentVolume func(pv *v1.PersistentVolume) (id string, relevant bool)
+	// MatchProvisioner evaluates if the StorageClass provisioner matches the running predicate
+	MatchProvisioner func(sc *storage.StorageClass) (relevant bool)
+	// IsMigrated returns a boolean specifying whether the plugin is migrated to a CSI driver
+	IsMigrated func(csiNode *storage.CSINode) bool
+}
+
+// EBSVolumeFilter is a VolumeFilter for filtering AWS ElasticBlockStore Volumes.
+var EBSVolumeFilter = VolumeFilter{
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: any reason why this is a global variable? can't we have it defined inside ```NewEBS()``` function?

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+
+const (
+	// EBSVolumeFilterType defines the filter name for EBSVolumeFilter.
+	EBSVolumeFilterType = "EBS"
+	// GCEPDVolumeFilterType defines the filter name for GCEPDVolumeFilter.
+	GCEPDVolumeFilterType = "GCE"
+	// AzureDiskVolumeFilterType defines the filter name for AzureDiskVolumeFilter.
+	AzureDiskVolumeFilterType = "AzureDisk"
+	// CinderVolumeFilterType defines the filter name for CinderVolumeFilter.
+	CinderVolumeFilterType = "Cinder"
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: make them local to the package and combine with the const block above.

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+const GCEPDName = "GCEPDLimits"
+
+// NewGCEPD returns function that initializes a new plugin and returns it.
+func NewGCEPD(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin, error) {
+	informerFactory := handle.SharedInformerFactory()
+	return NewNonCSILimitsWithInformerFactory(GCEPDVolumeFilterType, informerFactory), nil
+}
+
+// NonCSILimits contains information to check the max number of volumes for a plugin.
+type NonCSILimits struct {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: can we make it local to the package?

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+	// The string below is generated randomly during the struct's initialization.
+	// It is used to prefix volumeID generated inside the predicate() method to
+	// avoid conflicts with any real volume.
+	randomVolumeIDPrefix string
+}
+
+var _ framework.FilterPlugin = &NonCSILimits{}
+
+// NewNonCSILimitsWithInformerFactory returns a plugin with filter name and informer factory.
+func NewNonCSILimitsWithInformerFactory(
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: can we make it local?

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+// number of volumes which match a filter that it requests, and those that are already present.
+//
+// DEPRECATED
+// All cloudprovider specific predicates defined here are deprecated in favour of CSI volume limit
+// predicate - MaxCSIVolumeCountPred.
+//
+// The predicate looks for both volumes used directly, as well as PVC volumes that are backed by relevant volume
+// types, counts the number of unique volumes, and rejects the new pod if it would place the total count over
+// the maximum.
+func NewNonCSILimits(
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: can we make it local?

**Review by ahg-g — COMMENTED** [review]
Thanks

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+	// defaultMaxGCEPDVolumes defines the maximum number of PD Volumes for GCE.
+	// GCE instances can have up to 16 PD volumes attached.
+	defaultMaxGCEPDVolumes = 16
+	// defaultMaxAzureDiskVolumes defines the maximum number of PD Volumes for Azure.
+	// Larger Azure VMs can actually have much more disks attached.
+	// TODO We should determine the max based on VM size
+	defaultMaxAzureDiskVolumes = 16
+)
+
+const (
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: why two blocks of constants?

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+	scLister storagelisters.StorageClassLister,
+	pvLister corelisters.PersistentVolumeLister,
+	pvcLister corelisters.PersistentVolumeClaimLister,
+) framework.Plugin {
+	var filter VolumeFilter
+	var volumeLimitKey v1.ResourceName
+	var name string
+
+	switch filterName {
+
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: nit: seems like extra line

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+		volumeLimitKey = v1.ResourceName(volumeutil.AzureVolumeLimitKey)
+	case cinderVolumeFilterType:
+		name = CinderName
+		filter = cinderVolumeFilter
+		volumeLimitKey = v1.ResourceName(volumeutil.CinderVolumeLimitKey)
+	default:
+		klog.Fatalf("Wrong filterName, Only Support %v %v %v %v", ebsVolumeFilterType,
+			gcePDVolumeFilterType, azureDiskVolumeFilterType, cinderVolumeFilterType)
+		return nil
+
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: nit: seems like extra line

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+	node := nodeInfo.Node()
+	if node == nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("node not found"))
+	}
+
+	var (
+		csiNode *storage.CSINode
+		err     error
+	)
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: not sure this is easier to read than having the var for each variable, this is basically doubling the number of lines used for no reason.

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+	// filter out already-mounted volumes
+	for k := range existingVolumes {
+		if _, ok := newVolumes[k]; ok {
+			delete(newVolumes, k)
+		}
+	}
+
+	numNewVolumes := len(newVolumes)
+	maxAttachLimit := pl.maxVolumeFunc(node)
+
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: nit: seems like extra line

**Review by ahg-g — COMMENTED** [review]
looking at it again, I am not sure why the individual files of each provider was combined into one, it would have been more readable if they were separated (and add a file for the common code), but since you already combined them, I am not going to ask you to revert the change.

**Comment by draveness** [comment]
> looking at it again, I am not sure why the individual files of each provider was combined into one, it would have been more readable if they were separated (and add a file for the common code), but since you already combined them, I am not going to ask you to revert the change.

Thanks, I combined them because all of the codes in the `non_csi` file will be removed after the CSI migration graduates into GA.

**Comment by draveness** [comment]
I have addressed the comments, PTAL, Thanks!

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go (hunk 9 above)
Comment (draveness) [reply pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go]: done, this is a direct copy of the original code, and I prefer having the var for each of them too,

**Pull request merged**